src/imex.rs

```diff
@@ -106,6 +107,178 @@ pub async fn imex(
+        })
+        .await;
+
+    context.free_ongoing().await;
+
+    if let Err(err) = res.as_ref() {
+        // We are using Anyhow's .context() and to show the inner error, too, we need the {:#}:
+        error!(context, "IMEX failed to complete: {:#}", err);
+        context.emit_event(EventType::ImexProgress(0));
+    } else {
+        info!(context, "IMEX successfully completed");
+        context.emit_event(EventType::ImexProgress(1000));
+    }
+
+    res
+}
+
+pub async fn receive_backup_inner(
+    context: &Context,
+    ticket_bytes: Vec<u8>,
+    passphrase: String,
+) -> Result<()> {
+    use iroh_share::{Receiver, Ticket};
+    ensure!(
+        !context.is_configured().await?,
+        "Cannot import backups to accounts in use."
+    );
+    ensure!(
+        context.scheduler.read().await.is_none(),
+        "cannot import backup, IO is running"
+    );
+    let ticket = Ticket::from_bytes(&ticket_bytes)?;
+
+    let recv_dir = tempfile::tempdir().unwrap();
+    let recv_db = recv_dir.path().join("db");
+    let port = 9991;
+
+    let receiver = Receiver::new(port, &recv_db)
+        .await
+        .context("failed to create sender")?;
+    let mut receiver_transfer = receiver
+        .transfer_from_ticket(&ticket)
+        .await
+        .context("failed to read transfer")?;
+    let data = receiver_transfer.recv().await?;
+    let mut progress = receiver_transfer.progress()?;
+
+    context.sql.config_cache.write().await.clear();
+
+    // progress report
+    let ctx = context.clone();
+    let progress_task = tokio::spawn(async move {
+        let mut last_progress = 0;
+        while let Some(ev) = progress.next().await {
+            match ev {
+                Ok(iroh_share::ProgressEvent::Piece { index, total }) => {
+                    let progress = 1000 * index / total;
+                    if progress != last_progress && progress < 1000 {
+                        ctx.emit_event(EventType::ImexProgress(progress));
+                        last_progress = progress;
+                    }
+                }
+                Err(err) => {
+                    error!(ctx, "IMEX receive backup failed to complete: {}", err);
+                    ctx.emit_event(EventType::ImexProgress(0));
+                }
+            }
+        }
+    });
+
+    let out = context.get_blobdir();
+
+    let mut read_dir_stream = data
+        .read_dir()?
+        .ok_or_else(|| anyhow::anyhow!("unexpected data structure"))?;
+
+    while let Some(link) = read_dir_stream.next().await {
+        let link = link?;
+        let file_content = data.read_file(&link).await?;
+        let name = link.name.unwrap_or_default();
+        let path = out.join(&name);
+
+        let mut file = tokio::fs::File::create(&path)
+            .await
+            .with_context(|| format!("create file: {}", path.display()))?;
+        let mut content = file_content.pretty()?;
+        tokio::io::copy(&mut content, &mut file)
```

Comment:
`file` should probably be wrapped in a `tokio::io::BufWriter` (with a capacity of around 124kB for the average filesystems IIRC, mobile devices could be different though).